File: app/src/main/java/com/bassamalim/athkar/receivers/NotificationReceiver.java
package com.bassamalim.athkar.receivers;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.AudioManager;
import android.os.Build;
import android.util.Log;

import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.preference.PreferenceManager;

import com.bassamalim.athkar.Constants;
import com.bassamalim.athkar.R;
import com.bassamalim.athkar.Splash;
import com.bassamalim.athkar.views.AlathkarView;
import com.bassamalim.athkar.views.QuranView;

public class NotificationReceiver extends BroadcastReceiver {

    private Context context;
    private int id;
    private boolean isPrayer;
    private String channelId = "";
    private int type;
    private String action;
    private long time;

    @Override
    public void onReceive(Context gContext, Intent intent) {
        context = gContext;
        id = intent.getIntExtra("id", 10);
        action = intent.getAction();
        time = intent.getLongExtra("time", 0);

        if (PreferenceManager.getDefaultSharedPreferences(context).getInt(
                id+"notification_type", 2) != 0) {

            Log.i(Constants.TAG, "in notification receiver for " + id);

            publish();

            /*if (Build.VERSION.SDK_INT >= 26)
                context.startForegroundService(intent1);
            else
                context.startService(intent1);*/
        }
    }

    private void publish() {
        long max = time + 120000;
        if (System.currentTimeMillis() <= max) {
            isPrayer = action.equals("prayer");
            type = PreferenceManager.getDefaultSharedPreferences(context)
                    .getInt(id+"notification_type", 2);

            Log.i(Constants.TAG, "in notification service for " + id);

            createNotificationChannel();
            Notification notification = build();

            if (isPrayer) {
                AudioManager am = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
                // Request audio focus                                   // Request permanent focus.
                am.requestAudioFocus(null, AudioManager.STREAM_MUSIC,
                        AudioManager.AUDIOFOCUS_GAIN);
            }

            NotificationManagerCompat managerCompat = NotificationManagerCompat.from(context);
            if (id == 7)    // so that night notification would replace morning notification
                id--;
            managerCompat.notify(id, notification);

            //startForeground(NOTIFICATION_ID, notification);
        }
        else
            Log.e(Constants.TAG, "Late intent walking");
    }

    private Notification build() {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, channelId);
        builder.setSmallIcon(R.drawable.ic_athan);
        Bitmap icon = BitmapFactory.decodeResource(context.getResources(), R.drawable.ic_athan);
        builder.setLargeIcon(icon);
        builder.setTicker(context.getResources().getString(R.string.app_name));

        switch (id) {
            case 0: {
                builder.setContentTitle("صلاة الفجر");
                builder.setContentText("حان موعد أذان الفجر");
                break;
            }
            case 1: {
                builder.setContentTitle("الشروق");
                builder.setContentText("حان وقت الشروق");
                break;
            }
            case 2: {
                builder.setContentTitle("صلاة الظهر");
                builder.setContentText("حان موعد أذان الظهر");
                break;
            }
            case 3: {
                builder.setContentTitle("صلاة العصر");
                builder.setContentText("حان موعد أذان العصر");
                break;
            }
            case 4: {
                builder.setContentTitle("صلاة المغرب");
                builder.setContentText("حان موعد أذان المغرب");
                break;
            }
            case 5: {
                builder.setContentTitle("صلاة العشاء");
                builder.setContentText("حان موعد أذان العشاء");
                break;
            }
            case 6: {
                builder.setContentTitle("أذكار الصباح");
                builder.setContentText("اضغط لقراءة أذكار الصباح");
                break;
            }
            case 7: {
                builder.setContentTitle("أذكار المساء");
                builder.setContentText("اضغط لقراءة أذكار المساء");
                break;
            }
            case 8: {
                builder.setContentTitle("صفحة اليوم");
                builder.setContentText("اضغط لقراءة صفحة اليوم من المصحف");
                break;
            }
            case 9: {
                builder.setContentTitle("جمعة مباركة");
                builder.setContentText("اضغط لقراءة سورة الكهف");
                break;
            }
        }

        builder.setPriority(NotificationCompat.PRIORITY_MAX);
        builder.setAutoCancel(true);
        builder.setOnlyAlertOnce(true);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
            builder.setColor(context.getColor(R.color.secondary));
        else
            builder.setColor(context.getResources().getColor(R.color.secondary));
        builder.setContentIntent(onClick(id));

        if (type==1)
            builder.setSilent(true);

        //builder.setOngoing(true);
        //builder.setDeleteIntent(contentPendingIntent)  // if needed

        // NO_CLEAR makes the notification stay when the user performs a "delete all" command
        //notification.flags = Notification.FLAG_NO_CLEAR;
        return builder.build();
    }

    private PendingIntent onClick(int variable) {
        Intent intent;
        PendingIntent pendingIntent;

        if (variable == 6) {
            intent = new Intent(context, AlathkarView.class);
            intent.putExtra("category", 0);
            intent.putExtra("index", 0);
            intent.putExtra("title", "أذكار الصباح");
        }
        else if (variable == 7) {
            intent = new Intent(context, AlathkarView.class);
            intent.putExtra("category", 0);
            intent.putExtra("index", 1);
            intent.putExtra("title", "أذكار المساء");
        }
        else if (variable == 8) {
            intent = new Intent(context, QuranView.class);
            intent.setAction("random");
        }
        else if (variable == 9) {
            intent = new Intent(context, QuranView.class);
            intent.setAction("specific");
            intent.putExtra("surah_index", 17);    // alkahf
        }
        else
            intent = new Intent(context, Splash.class);

        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            pendingIntent = PendingIntent.getActivity(context, variable, intent,
                    PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
        }
        else {
            pendingIntent = PendingIntent.getActivity(context,
                    variable, intent, PendingIntent.FLAG_UPDATE_CURRENT);
        }

        return pendingIntent;
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CharSequence name = "";
            String description = "";
            if (isPrayer) {
                channelId = "Athan";
                name = "إشعارات الصلوات";
            }
            else if (id == 1) {
                channelId = "sunrise";
                name = "إشعار الشروق";
            }
            else if (id == 6 || id == 7) {
                channelId = "morning and night";
                name = "إشعارات أذكار الصباح والمساء";
            }
            else if (id == 8) {
                channelId = "daily_page";
                name = "إشعار صفحة اليوم";
            }
            else if (id == 9) {
                channelId = "friday_kahf";
                name = "إشعار سورة الكهف";
            }
            int importance = NotificationManager.IMPORTANCE_HIGH;
            NotificationChannel notificationChannel  = new NotificationChannel(
                    channelId, name, importance);
            notificationChannel.setDescription(description);
            notificationChannel.setLockscreenVisibility(Notification.VISIBILITY_PUBLIC);
            NotificationManager notificationManager = context.getSystemService(NotificationManager.class);
            notificationManager.createNotificationChannel(notificationChannel);
        }
    }

}

File: app/src/main/java/bassamalim/hidaya/core/receivers/NotificationReceiver.kt
package bassamalim.hidaya.core.receivers

import android.Manifest
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.os.Build
import android.util.Log
import androidx.compose.ui.graphics.toArgb
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import bassamalim.hidaya.R
import bassamalim.hidaya.core.Activity
import bassamalim.hidaya.core.Globals
import bassamalim.hidaya.core.data.repositories.AppSettingsRepository
import bassamalim.hidaya.core.data.repositories.NotificationsRepository
import bassamalim.hidaya.core.data.repositories.QuranRepository
import bassamalim.hidaya.core.enums.NotificationType
import bassamalim.hidaya.core.enums.Reminder
import bassamalim.hidaya.core.enums.ThemeColor
import bassamalim.hidaya.core.nav.Screen
import bassamalim.hidaya.core.services.AthanService
import bassamalim.hidaya.core.ui.theme.getThemeColor
import bassamalim.hidaya.core.utils.ActivityUtils
import bassamalim.hidaya.features.quran.reader.domain.QuranTarget
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import java.util.Calendar
import javax.inject.Inject
import kotlin.math.abs

@AndroidEntryPoint
class NotificationReceiver : BroadcastReceiver() {

    private lateinit var ctx: Context
    @Inject lateinit var notificationsRepository: NotificationsRepository
    @Inject lateinit var quranRepository: QuranRepository
    @Inject lateinit var appSettingsRepository: AppSettingsRepository
    private var notificationId = 0
    private var channelId = ""

    @OptIn(DelicateCoroutinesApi::class)
    override fun onReceive(context: Context, intent: Intent) {
        ctx = context.applicationContext

        val reminder = Reminder.getById(intent.getIntExtra("id", -1))
        Log.d(Globals.TAG, "notification receiver: received $reminder")
        val time = intent.getLongExtra("time", 0L)
        notificationId = reminder.id

        GlobalScope.launch {
            ActivityUtils.configure(
                context = context,
                applicationContext = context.applicationContext,
                language = appSettingsRepository.getLanguage().first()
            )

            if (!isOnTime(time) || isAlreadyNotified(reminder)) {
                Log.i(Globals.TAG, "notification receiver: not on time or already notified")
                return@launch
            }

            when (reminder) {
                is Reminder.Prayer -> handlePrayerReminder(reminder, time)
                is Reminder.PrayerExtra -> handlePrayerExtraReminder(reminder)
                is Reminder.Devotional -> handleDevotionalReminder(reminder)
            }
        }
    }

    private suspend fun handlePrayerReminder(reminder: Reminder.Prayer, time: Long) {
        Log.i(Globals.TAG, "in notification receiver for $reminder prayer")

        val notificationType = notificationsRepository.getNotificationType(reminder).first()

        if (notificationType != NotificationType.OFF) {
            if (notificationType == NotificationType.ATHAN) startService(reminder, time)
            else showNotification(reminder, notificationType)
        }
    }

    private suspend fun handlePrayerExtraReminder(reminder: Reminder.PrayerExtra) {
        Log.i(Globals.TAG, "in notification receiver for $reminder reminder")

        showNotification(reminder)
    }

    private suspend fun handleDevotionalReminder(reminder: Reminder.Devotional) {
        Log.i(Globals.TAG, "in notification receiver for $reminder extra")

        showNotification(reminder)
    }

    private fun isOnTime(time: Long): Boolean {
        val max = time + 120000
        return System.currentTimeMillis() <= max
    }

    private suspend fun isAlreadyNotified(reminder: Reminder): Boolean {
        val lastDate = notificationsRepository.getLastNotificationDates().first()[reminder]
        return lastDate == Calendar.getInstance()[Calendar.DAY_OF_YEAR]
    }

    private suspend fun showNotification(
        reminder: Reminder,
        notificationType: NotificationType = NotificationType.NOTIFICATION
    ) {
        createNotificationChannel(reminder)

        if (reminder is Reminder.Prayer && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val audioManager = ctx.getSystemService(Service.AUDIO_SERVICE) as AudioManager
            // Request audio focus
            audioManager.requestAudioFocus(                   // Request permanent focus
                AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN)
                    .setAudioAttributes(
                        AudioAttributes.Builder()
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .setUsage(AudioAttributes.USAGE_ALARM)
                            .build()
                    ).build()
            )
        }

        if (ActivityCompat.checkSelfPermission(
                ctx, Manifest.permission.POST_NOTIFICATIONS
            ) == PackageManager.PERMISSION_GRANTED) {
            val notification = build(reminder, notificationType)
            NotificationManagerCompat.from(ctx).notify(notificationId, notification)
        }

        markAsNotified(reminder)
    }

    private fun startService(reminder: Reminder.Prayer, time: Long) {
        val intent = Intent(ctx, AthanService::class.java).apply {
            action = Globals.PLAY_ATHAN
            putExtra("id", reminder.id)
            putExtra("time", time)
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            ctx.startForegroundService(intent)
        else
            ctx.startService(intent)

        markAsNotified(reminder)
    }

    private suspend fun build(
        reminder: Reminder,
        notificationType: NotificationType
    ): Notification {
        return NotificationCompat.Builder(ctx, channelId).apply {
            setSmallIcon(R.drawable.ic_athan)
            setTicker(ctx.resources.getString(R.string.app_name))

            setContentTitle(getTitle(reminder))
            setContentText(getSubtitle(reminder))

            priority = NotificationCompat.PRIORITY_MAX
            setAutoCancel(true)
            setOnlyAlertOnce(true)
            this.color = getThemeColor(
                color = ThemeColor.SURFACE_CONTAINER,
                theme = appSettingsRepository.getTheme().first()
            ).toArgb()
            setContentIntent(onClick(reminder))

            if (notificationType == NotificationType.SILENT)
                setSilent(true)
        }.build()
    }

    private fun getTitle(reminder: Reminder): String {
        return if ((reminder == Reminder.Prayer.Dhuhr || reminder == Reminder.PrayerExtra.Dhuhr) &&
            Calendar.getInstance()[Calendar.DAY_OF_WEEK] == Calendar.FRIDAY) {
            if (reminder is Reminder.PrayerExtra)
                ctx.resources.getString(R.string.jumuah_reminder_title)
            else
                ctx.resources.getString(R.string.jumuah_title)
        }
        else ctx.resources.getStringArray(R.array.prayer_titles)[reminder.id-1]
    }

    private suspend fun getSubtitle(reminder: Reminder): String {
        return if (reminder is Reminder.PrayerExtra) {
            val offset = notificationsRepository.getPrayerExtraReminderTimeOffsets()
                .first()[reminder]!!
            String.format(
                format =
                    if (offset < 0) ctx.resources.getString(R.string.reminder_before)
                    else ctx.resources.getString(R.string.reminder_after),
                    if (reminder == Reminder.PrayerExtra.Dhuhr &&
                        Calendar.getInstance()[Calendar.DAY_OF_WEEK] == Calendar.FRIDAY)
                        ctx.resources.getString(R.string.jumuah)
                    else
                        ctx.resources.getStringArray(R.array.prayer_names)
                            [reminder.toPrayer().toReminder().id-1],
                abs(offset) // to remove - sign
            )
        }
        else {
            if (reminder == Reminder.Prayer.Dhuhr &&
                Calendar.getInstance()[Calendar.DAY_OF_WEEK] == Calendar.FRIDAY)
                ctx.resources.getString(R.string.jumuah_subtitle)
            else ctx.resources.getStringArray(R.array.prayer_subtitles)[reminder.id-1]
        }
    }

    private suspend fun onClick(reminder: Reminder): PendingIntent {
        val intent = Intent(ctx, Activity::class.java)

        val route = when (reminder) {
            Reminder.Devotional.MorningRemembrances -> Screen.RemembranceReader(0.toString()).route
            Reminder.Devotional.EveningRemembrances -> Screen.RemembranceReader(1.toString()).route
            Reminder.Devotional.DailyWerd -> {
                Screen.QuranReader(
                    targetType = QuranTarget.PAGE.name,
                    targetValue = quranRepository.getWerdPageNum().first().toString()
                ).route
            }
            Reminder.Devotional.FridayKahf -> {
                Screen.QuranReader(
                    targetType = QuranTarget.SURA.name,
                    targetValue = 17.toString() // surat al-kahf
                ).route
            }
            else -> Screen.Main.route
        }
        intent.putExtra("start_route", route)

        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK

        return PendingIntent.getActivity(
            ctx,
            reminder.id,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
    }

    private fun createNotificationChannel(reminder: Reminder) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            channelId = getChannelId(reminder)
            val channel = NotificationChannel(
                channelId,
                getChannelName(reminder),
                NotificationManager.IMPORTANCE_HIGH
            )
            channel.description = "description"
            channel.lockscreenVisibility = Notification.VISIBILITY_PUBLIC
            val notificationManager = ctx.getSystemService(NotificationManager::class.java)
            notificationManager.createNotificationChannel(channel)
        }
    }

    private fun getChannelId(reminder: Reminder): String {
        return when (reminder) {
            is Reminder.Prayer -> "prayers_channel"
            is Reminder.PrayerExtra -> "prayers_extra_channel"
            is Reminder.Devotional -> {
                when (reminder) {
                    Reminder.Devotional.MorningRemembrances,
                    Reminder.Devotional.EveningRemembrances ->
                        "morning_and_evening_remembrances_channel"
                    Reminder.Devotional.DailyWerd -> "daily_werd_channel"
                    Reminder.Devotional.FridayKahf -> "friday_kahf_channel"
                }
            }
        }
    }

    private fun getChannelName(reminder: Reminder): String {
        return when (reminder) {
            is Reminder.Prayer -> ctx.getString(R.string.prayer_notification_channel)
            is Reminder.PrayerExtra -> ctx.getString(R.string.prayer_extra_notification_channel)
            is Reminder.Devotional -> {
                when (reminder) {
                    Reminder.Devotional.MorningRemembrances,
                    Reminder.Devotional.EveningRemembrances -> ctx.getString(
                        R.string.morning_and_evening_remembrances_notification_channel
                    )
                    Reminder.Devotional.DailyWerd -> ctx.getString(
                        R.string.daily_werd_notification_channel
                    )
                    Reminder.Devotional.FridayKahf -> ctx.getString(
                        R.string.friday_kahf_notification_channel
                    )
                }
            }
        }
    }

    private fun markAsNotified(reminder: Reminder) {
        notificationsRepository.setLastNotificationDate(
            reminder = reminder,
            dayOfYear = Calendar.getInstance()[Calendar.DAY_OF_YEAR]
        )
    }

}